File: 101-200/111-120/113-pathSum2/pathSum2.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum, path=[]):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        
        if not root:
            return []
        if not root.left and not root.right:
            return [path + [root.val]] if sum == root.val else []
        
        return self.pathSum(root.left, sum-root.val, path+[root.val]) + self.pathSum(root.right, sum-root.val, path+[root.val])

File: 101-200/111-120/113-pathSum2/pathSum2.cpp
#include <vector>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

using namespace std;

class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<int> path;
        vector<vector<int>> result;
        pathSumHelper(root, sum, path, result);
        return result;
    }

    void pathSumHelper(TreeNode* node, int sum, vector<int> &path, vector<vector<int>> &result) {
        if (!node)
            return;

        path.push_back(node->val);
        if (!node->left && !node->right && sum == node->val)
            result.push_back(path);
        else {
            pathSumHelper(node->left, sum-node->val, path, result);
            pathSumHelper(node->right, sum-node->val, path, result);
        }

        path.pop_back();
    }
};